File: 0216-combination-sum-iii/0216-combination-sum-iii.cpp
class Solution {
public:
void solve(vector<int>& temp,vector<vector<int>>& results,vector<int>& candidates,int index,int target,int n,int k) {
    if(target==0 && temp.size()==k) {
        results.push_back(temp);
        return;
    } else if(target==0) { return; }
    for(int i=index;i<n;i++) {
         if(i!=index && candidates[i]==candidates[i-1]) {
                //Remove the duplicate combination case
                continue;
            }
        if(target-candidates[i]>=0) {
           
            temp.push_back(candidates[i]);
            solve(temp,results,candidates,i+1,target-candidates[i],n,k);
            temp.pop_back();
        }
    }
    
    
}
    vector<vector<int>> combinationSum3(int k,int n) {
        vector<vector<int>> results;
        vector<int> temp;
        vector<int> candidates;
        for(int i=1;i<=9;i++) candidates.push_back(i);
        solve(temp,results,candidates,0,n,candidates.size(),k);
        return results;

    }
};

File: 0216-combination-sum-iii/0216-combination-sum-iii.py
class Solution:
    def combinationSum3(self, k: int, target: int) -> List[List[int]]:
        candidates = [x for x in range(1,10)]
        answer = []
        curr_answer = []
        candidates.sort()
        def solution(index,new_target,curr_answer,k_left):
            if new_target<0 or k_left<0: return
            if new_target==0 and k_left==0:
                answer.append(curr_answer.copy())
                return
            #pick or no pick
            for i,num in enumerate(candidates[index:]):
                #pick
                if i>0 and candidates[i+index]==candidates[i-1+index]: continue
                curr_answer.append(num)
                solution(index+i+1,new_target-num,curr_answer,k_left-1)
                curr_answer.pop()
        solution(0,target,curr_answer,k)
        return answer

    